File: src/audit.ts
import path from "node:path";
import { appendPrivateJsonl, DEFAULT_STATE_DIR } from "./state.js";
import { redactCommand } from "./security-policy.js";

export type AuditEvent = {
  ts?: string;
  event: string;
  requestId?: string;
  workspaceKey?: string;
  tool?: string;
  processId?: string;
  approvalId?: string;
  status?: string;
  riskLevel?: string;
  detail?: unknown;
};

const AUDIT_ENABLED = process.env.CODELOCAL_AUDIT_FILE !== "0";
const AUDIT_FILE = process.env.CODELOCAL_AUDIT_PATH ?? path.join(DEFAULT_STATE_DIR, "audit.jsonl");

function sanitize(value: unknown, key = ""): unknown {
  if (value instanceof Error) return { name: value.name, message: value.message };
  if (Array.isArray(value)) return value.map((item) => sanitize(item));
  if (typeof value === "string") {
    if (/command|detail|query/i.test(key)) return redactCommand(value).slice(0, 2000);
    if (/content|patch|input|oldText|newText/i.test(key)) return `[${Buffer.byteLength(value, "utf8")} bytes]`;
    return value.slice(0, 4000);
  }
  if (!value || typeof value !== "object") return value;
  const out: Record<string, unknown> = {};
  for (const [childKey, childValue] of Object.entries(value as Record<string, unknown>)) {
    if (/token|secret|password|authorization|cookie|credentialSecret/i.test(childKey)) out[childKey] = "[REDACTED]";
    else out[childKey] = sanitize(childValue, childKey);
  }
  return out;
}

export async function audit(event: AuditEvent) {
  if (!AUDIT_ENABLED) return;
  const record = sanitize({ ts: event.ts ?? new Date().toISOString(), ...event }) as Record<string, unknown>;
  await appendPrivateJsonl(AUDIT_FILE, record).catch(() => undefined);
}

File: internal/audit/audit.go
package audit

import (
	"fmt"
	"os"
	"path/filepath"
	"reflect"
	"regexp"
	"time"

	"github.com/0xmarkhydra/codelocal/internal/security"
	"github.com/0xmarkhydra/codelocal/internal/state"
)

type Event struct {
	TS           string `json:"ts,omitempty"`
	Event        string `json:"event"`
	RequestID    string `json:"requestId,omitempty"`
	MCPSessionID string `json:"mcpSessionId,omitempty"`
	WorkspaceKey string `json:"workspaceKey,omitempty"`
	Tool         string `json:"tool,omitempty"`
	ProcessID    string `json:"processId,omitempty"`
	ApprovalID   string `json:"approvalId,omitempty"`
	Status       string `json:"status,omitempty"`
	RiskLevel    string `json:"riskLevel,omitempty"`
	Detail       any    `json:"detail,omitempty"`
}

var secretKeyRE = regexp.MustCompile(`(?i)(?:api[_-]?key|access[_-]?key|private[_-]?key|token|secret|password|passwd|authorization|cookie|credentialSecret)`)
var redactTextKeyRE = regexp.MustCompile(`(?i)command|detail|query`)

// Free-form input text can contain passwords, OTPs, messages, form values or
// other private content. Audit the presence/size of such payloads, never the
// payload itself.
var sizeOnlyKeyRE = regexp.MustCompile(`(?i)^(?:content|patch|input|text|oldText|newText)$`)

func Enabled() bool { return os.Getenv("CODELOCAL_AUDIT_FILE") != "0" }

func Path() string {
	if value := os.Getenv("CODELOCAL_AUDIT_PATH"); value != "" {
		return value
	}
	return filepath.Join(state.Dir(), "audit.jsonl")
}

func sanitize(value any, key string) any {
	if value == nil {
		return nil
	}
	if err, ok := value.(error); ok {
		return map[string]any{"name": reflect.TypeOf(err).String(), "message": sanitize(err.Error(), "detail")}
	}
	switch v := value.(type) {
	case string:
		if sizeOnlyKeyRE.MatchString(key) {
			return fmt.Sprintf("[%d bytes]", len([]byte(v)))
		}
		if redactTextKeyRE.MatchString(key) {
			v = security.RedactCommand(v)
			if len(v) > 2000 {
				v = v[:2000]
			}
			return v
		}
		if len(v) > 4000 {
			return v[:4000]
		}
		return v
	case []string:
		out := make([]any, len(v))
		for i := range v {
			out[i] = sanitize(v[i], key)
		}
		return out
	case []any:
		out := make([]any, len(v))
		for i := range v {
			out[i] = sanitize(v[i], key)
		}
		return out
	case map[string]any:
		out := map[string]any{}
		for childKey, child := range v {
			if secretKeyRE.MatchString(childKey) {
				out[childKey] = "[REDACTED]"
			} else {
				out[childKey] = sanitize(child, childKey)
			}
		}
		return out
	default:
		return value
	}
}

func Write(event Event) {
	if !Enabled() {
		return
	}
	if event.TS == "" {
		event.TS = time.Now().UTC().Format(time.RFC3339Nano)
	}
	record := map[string]any{
		"ts": event.TS, "event": event.Event, "requestId": event.RequestID, "mcpSessionId": event.MCPSessionID,
		"workspaceKey": event.WorkspaceKey, "tool": event.Tool, "processId": event.ProcessID, "approvalId": event.ApprovalID,
		"status": event.Status, "riskLevel": event.RiskLevel, "detail": sanitize(event.Detail, "detail"),
	}
	for key, value := range record {
		if value == "" || value == nil {
			delete(record, key)
		}
	}
	_ = state.AppendJSONL(Path(), record)
}
